specs/BUNDLE_SPEC.md

```diff
@@ -0,0 +1,271 @@
+image.
+
+Assuming there are artifacts present, one of the two above calls will return an
+image index listing the artifacts which have been associated with the specified
+image:
+
+```
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "digest": "sha256:badf00d..",
+      "size": 779,
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2"
+    }
+  ]
+}
+```
+
+From this the client can identify any Sigstore bundles by looking at the
+`artifactType` field.
+
+Using the `digest` listed in the image index, the next step is to retrieve the
+manifest for the bundle:
+
+```
+GET /v2/foo/manifests/sha256:badf00d..
+```
+
+```
+{
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "schemaVersion": 2,
+  "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+  "config": {
+    "mediaType": "application/vnd.oci.empty.v1+json",
+    "digest": "sha256:44136fa3...",
+    "size": 2
+  },
+  "layers": [
+    {
+      "digest": "sha256:cafed00d...",
+      "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "size": 4971
+    }
+  ],
+  "subject": {
+    "digest": "sha256:c00010ff...",
+    "mediaType": "application/vnd.oci.image.index.v1+json"
+   }
+}
+```
+
+The final step is to use the `digest` from the first of the `layers` to retrieve
+the bundle blob:
+
+```
+GET /v2/foo/blobs/uploads/?digest=cafed00d...
+```
+
+```
+{
+  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+  "verificationMaterial": {...},
+  "messageSignature": {...}
+}
+```
+
+## Annotations
+
+For any given image, there may be any number of attached attestation bundles.
+When there are multiple Sigstore bundles associated with an image it may be
+difficult to identify which artifact is which in the image index:
+
+```json
+{
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "schemaVersion": 2,
+  "manifests": [
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "digest": "sha256:facefeed",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    },
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "digest": "sha256:d0d0caca",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    }
+  ]
+}
+```
+
+To help disambiguate attestations, clients may add annotations to the items
+in the `manifests` list which indicate what is contained within each bundle:
+
+- `dev.sigstore.cosign.bundle.content` - Must be one "message-signature" or
+  "dsse-envelope" and should match the type of content embedded in the Sigstore
+  bundle.
+- `dev.sigstore.cosign.bundle.predicateType` - When the bundle contains a
+  DSSE-wrapped in-toto statement, the statement's predicate can be reflected
+  here.
+
+```json
```

Comment:
Just jotting this down here. Wondering if it would be useful to have a an example out there demonstrating a fully compliant spec?

Reply:
I don't know that I want to reference this in the spec, but I do have an example at `index.docker.io/bdehamer/hello:latest`.

You can poke at this with the `oras` CLI. Look-up an referring artifacts:

```
oras discover index.docker.io/bdehamer/hello:latest
```
```
Discovered 1 artifact referencing latest
Digest: sha256:01b2325c7cae9939e4484061c37d36e0b95fb3f5e66f80ff924582ba5939e831

Artifact Type                                   Digest
application/vnd.dev.sigstore.bundle.v3.0+json   sha256:30bb112189b0070d8c440fb0c9ef13d4ff25014ccc64d90e5839f90b99c81779
```

Fetch the bundle manifest:
```
oras manifest fetch index.docker.io/bdehamer/hello@sha256:30bb112189b0070d8c440fb0c9ef13d4ff25014ccc64d90e5839f90b99c81779
```
```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.dev.sigstore.bundle.v3.0+json",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30="
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar",
      "digest": "sha256:0e6245a3f020384b3b10aa2f4fe838ce5e87a71038e710938875cc22d6adb917",
      "size": 10097,
      "annotations": {
        "org.opencontainers.image.title": "bundle.json"
      }
    }
  ],
  "subject": {
    "mediaType": "application/vnd.oci.image.index.v1+json",
    "digest": "sha256:01b2325c7cae9939e4484061c37d36e0b95fb3f5e66f80ff924582ba5939e831",
    "size": 855
  },
  "annotations": {
    "dev.sigstore.bundle.predicateType": "https://slsa.dev/provenance/v1",
    "dev.sigstore.cosign.bundle.content": "dsse-envelope",
    "org.opencontainers.image.created": "2024-03-28T19:44:17Z"
  }
}
```
Fetch the bundle:
```
oras blob fetch index.docker.io/bdehamer/hello@sha256:0e6245a3f020384b3b10aa2f4fe838ce5e87a71038e710938875cc22d6adb917 --output -
```
```
{"mediaType":"application/vnd.dev.sigstore.bundle.v3.0+json"...}
```